core/src/test/resources/sql/agg.iq

```diff
@@ -3467,4 +3467,38 @@ order by ename, deptno;
 
 !ok
 
+# Test cases for [CALCITE-5209] ArrayIndexOutOfBoundsException during SqlToRelConverter for group-by on `case`
+# having `in` expression predicates exceeding SqlRelConverter.Config InSubQueryThreshold
+!use scott
+select
+    case when deptno in (1, 2, 3, 4, 5) THEN 1 else 0 end
+from emp
+group by
+    case when deptno in (1, 2, 3, 4, 5) THEN 1 else 0 end;
+
++--------+
+| EXPR$0 |
++--------+
+|      0 |
++--------+
+(1 row)
+
+!ok
+
+!set insubquerythreshold 5
+select
+    case when deptno in (1, 2, 3, 4, 5) THEN 1 else 0 end
```

Comment:
The exception has nothing to do with `case`, I would suggest to make it simpler (at least for the commit message and Jira title)

Reply:
What about "Proper sub-query handling if it is used inside select list and group by" ?

Reply:
This looks better!

Reply:
@libenchao Done. I also rebased branch.